### PR TITLE
MSLI-4 Series details

### PR DESCRIPTION
**Pull request by IvanGit2021**
Open Details screen that loads information from API if Internet is available 

Loads cached data from Database if there is no access to API

Show a button to add Series to Series List or remove if already added (checkmark)

### CODE REVIEW

File: My Series List/My Series List/Model/Repository/SeriesRemoteDataSource.swift [hunk 1]
```diff
@@ -27,4 +29,20 @@ class SeriesRemoteDataSource {
         })
         task.resume()
     }
+    
+    func getDetails(id: Int32, completionHandler: @escaping (Result<Details, Error>) -> Void) {
+        let task = URLSession.shared.dataTask(with: URL (string: "\(detailsApiUrl)\(id)?\(detailsApiParameters)" )! , completionHandler: { data, response, error in
```
Comment (andre-exd) [inline My Series List/My Series List/Model/Repository/SeriesRemoteDataSource.swift]: Can you create a url variable and pass here instead of doing everything in a single line?

File: My Series List/My Series List/Model/SeriesDetailsModel.swift [hunk 2]
```diff
@@ -0,0 +1,36 @@
+//
+//  SeriesDetailsModel.swift
+//  My Series List
+//
+//  Created by Ivan Dias on 04/01/2022.
+//
+
+import Foundation
+
+class Details: Codable {
```
Comment (andre-exd) [inline My Series List/My Series List/Model/SeriesDetailsModel.swift]: Can you try to reuse Series class by adding missing fields there?

File: My Series List/My Series List/Model/SeriesModel.swift [hunk 3]
```diff
@@ -14,19 +14,17 @@ enum Api {
     }
     
     class Series: Codable {
-        let id: Double?
-        let title: String?
-        let overView: String?
-        let posterPath: String?
-        let voteAverage: Double?
+        var id: Int32?
```
Comment (andre-exd) [inline My Series List/My Series List/Model/SeriesModel.swift:21]: Don't understand why this change was done. Can you explain?

File: My Series List/My Series List/View/DetailsView/DetailsController.swift [hunk 4]
```diff
@@ -6,35 +6,117 @@
...
     
-    func startLoading() {
-        print("Start Loading")
+    func listDetails(details: Details) {
+        self.details = details
+        DispatchQueue.main.async { [self] in
+            emptyDetails.isHidden = true
+            let imageUrl = URL(string: details.backdropPath == nil ? "" : "https://image.tmdb.org/t/p/w500" + details.backdropPath!)
+            let imageUrl2 = URL(string: details.posterPath == nil ? "" : "https://image.tmdb.org/t/p/w500" + details.posterPath!)
+            if isSaved!{
```
Comment (andre-exd) [inline My Series List/My Series List/View/DetailsView/DetailsController.swift:56]: Format code

File: My Series List/My Series List/View/DetailsView/DetailsController.swift [hunk 5]
```diff
@@ -6,35 +6,117 @@
...
+            detailsOverview.isHidden = true
+            detailsWebSite.isHidden = true
+            detailsFavourites.isHidden = true
+            detailsSaveDelete.isHidden = true
         }
     }
     
-    func listEmpty() {
-        print("No data found")
+    func setSavedToFalse() {
```
Comment (andre-exd) [inline My Series List/My Series List/View/DetailsView/DetailsController.swift]: Can you consider a method with a different name? setSavedToFalse() should only change saved value as false.
updateFavorites() might be a better name. We should talk offline about this

File: My Series List/My Series List/View/DetailsView/DetailsPresenter.swift [hunk 6]
```diff
@@ -9,27 +9,79 @@ import Foundation
...
+                    break
+                case .success(let series):
+                    for series in series {
+                        if series.id == self.id {
+                            self.seriesRepository.deleteSeries(series: series) { results in
+                            }
+                            break
+                        }
+                    }
+                    self.detailsView!.setSavedToFalse()
```
Comment (andre-exd) [inline My Series List/My Series List/View/DetailsView/DetailsPresenter.swift]: Seems like the usage of setSavedToFalse is actually to update the view after toggling if a Series is stored or not. Please consider using something like updateFavoriteIcon() or something similar. Also, there should be not reason why that would fetch again the data from database or api instead of trying to change isSaved flag on item of list.
We can talk about this offline

File: My Series List/My Series List/View/DetailsView/DetailsPresenter.swift [hunk 7]
```diff
@@ -9,27 +9,79 @@ import Foundation
...
+        seriesRepository.searchSeriesDetails(id: id) { [self] results in
+            switch results {
+            case .success(let details):
+                self.details = details
+                detailsView?.listDetails(details: details)
+            case .failure(_):
+                DispatchQueue.main.async {
+                    seriesRepository.getSeries { results in
+                        switch results {
+                        case (.failure(_)):
```
Comment (andre-exd) [inline My Series List/My Series List/View/DetailsView/DetailsPresenter.swift]: Consider moving failure after handling success

File: My Series List/My Series List/View/DetailsView/DetailsPresenter.swift [hunk 8]
```diff
@@ -9,27 +9,79 @@ import Foundation
...
+                            break
+                        case (.success(let seriesCoreData)):
+                        var count = 0
```
Comment (andre-exd) [inline My Series List/My Series List/View/DetailsView/DetailsPresenter.swift]: Was not able to follow this logic. Can you move this code into a method and giving it a proper name.
In my understanding, you are searching in all Series in Database if you can find a match. This should be fixed by adding a new method to Repository called getSeriesById() that only return a Series

**Review by andre-exd — CHANGES_REQUESTED** [review]

File: My Series List/My Series List/Model/SeriesModel.swift (hunk 3 above)
Comment (IvanGit2021) [reply My Series List/My Series List/Model/SeriesModel.swift:21]: i changed this because I needed to instantiate a Series from this class and inject some data in it and it didn't allow me because the properties were constants

File: My Series List/My Series List/View/DetailsView/DetailsPresenter.swift [hunk 9]
```diff
@@ -9,27 +9,61 @@ import Foundation
...
 
 class DetailsPresenter {
     
     var detailsView: DetailsView?
     let seriesRepository = SeriesRepository()
-}
+    var id: Int32?
+    var details: Api.Series?
+   
+    func insertRemove(isSaved: Bool) {
```
Comment (andre-exd) [inline My Series List/My Series List/View/DetailsView/DetailsPresenter.swift]: Can you split this method into deleteSeries() and insertSeries()?

**Review by andre-exd — CHANGES_REQUESTED** [review]

**Review by andre-exd — APPROVED** [review]
